core/chains/evm/txmgr/evm_tx_store.go

```diff
@@ -440,18 +441,38 @@ func (o *evmTxStore) Transactions(offset, limit int) (txs []Tx, count int, err e
 	return
 }
 
+type TransactionsWithAttemptsSelector struct {
+	IdempotencyKey *string
+	Offset         uint64
+	Limit          uint64
+}
+
 // TransactionsWithAttempts returns all eth transactions with at least one attempt
 // limited by passed parameters. Attempts are sorted by id.
-func (o *evmTxStore) TransactionsWithAttempts(offset, limit int) (txs []Tx, count int, err error) {
-	sql := `SELECT count(*) FROM evm.txes WHERE id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts)`
-	if err = o.q.Get(&count, sql); err != nil {
-		return
+func (o *evmTxStore) TransactionsWithAttempts(selector TransactionsWithAttemptsSelector) (txs []Tx, count int, err error) {
+	stmt := sq.Select("count(*)").From("evm.txes").
+		Where("id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts)").PlaceholderFormat(sq.Dollar)
+	if selector.IdempotencyKey != nil {
+		stmt = stmt.Where("idempotency_key = ?", *selector.IdempotencyKey)
 	}
 
-	sql = `SELECT * FROM evm.txes WHERE id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts) ORDER BY id desc LIMIT $1 OFFSET $2`
+	query, args, err := stmt.ToSql()
+	if err != nil {
+		return nil, 0, fmt.Errorf("failed to build count query: %w", err)
+	}
+
+	if err = o.q.Get(&count, query, args...); err != nil {
+		return nil, 0, fmt.Errorf("failed to exec count query: %w, sql: %s", err, query)
+	}
+
+	query, args, err = stmt.RemoveColumns().Columns("*").
+		OrderBy("id desc").Limit(selector.Limit).Offset(selector.Offset).ToSql()
```

Comment:
@samsondav @reductionista WDYT about this github.com/Masterminds/squirrel library? It certainly helps in dynamic cases, but I'm worried about readability and efficiency.

Reply:
In my experience, SQL generators usually make things harder to read and understand than writing raw SQL. They sound nice in theory but don't work well in practice. Just my 2c

Reply:
@jmank88 @samsondav for this case, do you prefer strings builder/concatenation to handle the optional `idempotency_key`?

Reply:
I think it would be worth writing out to compare side-by-side at least.

Reply:
@jmank88 @samsondav 
```go
func (o *evmTxStore) TransactionsWithAttempts(selector TransactionsWithAttemptsSelector) (txs []Tx, count int, err error) {
	countQuery := "SELECT count(*) FROM evm.txes WHERE id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts)"
	selectQuery := "SELECT * FROM evm.txes WHERE id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts)"
	var args []interface{}
	if selector.IdempotencyKey != nil {
		args = append(args, *selector.IdempotencyKey)
		// NOTE: we are lucky to have only one optional filter and WHERE is the root statement, so we can use AND 
		selectQuery += " AND idempotency_key = ?"
		countQuery += " AND idempotency_key = ?"
	}

	countQuery = replacePositionalPlaceholders(countQuery)
	if err = o.q.Get(&count, countQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("failed to exec count query: %w, sql: %s", err, countQuery)
	}

	args = append(args, selector.Limit, selector.Offset)
	selectQuery += " ORDER BY id desc LIMIT ? OFFSET ?"
	var dbTxs []DbEthTx
	selectQuery = replacePositionalPlaceholders(selectQuery)
	if err = o.q.Select(&dbTxs, selectQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("failed to exec select query: %w, sql: %s", err, selectQuery)
	}
	txs = dbEthTxsToEvmEthTxs(dbTxs)
	err = o.preloadTxAttempts(txs)
	return
}

// Copy-paste from squirrel to replace ? with $1
func replacePositionalPlaceholders(sql string) string {
	buf := &bytes.Buffer{}
	i := 0
	for {
		p := strings.Index(sql, "?")
		if p == -1 {
			break
		}

		if len(sql[p:]) > 1 && sql[p:p+2] == "??" { // escape ?? => ?
			buf.WriteString(sql[:p])
			buf.WriteString("?")
			if len(sql[p:]) == 1 {
				break
			}
			sql = sql[p+2:]
		} else {
			i++
			buf.WriteString(sql[:p])
			fmt.Fprintf(buf, "$%d", i)
			sql = sql[p+1:]
		}
	}

	buf.WriteString(sql)
	return buf.String()
}
```

Reply:
I was not expecting the additional complexity of `replacePositionalPlaceholders`. Why is that necessary?
Otherwise, I think we can simplify the rest of the logic a little more :+1:

Reply:
It would be great if we could get to something like this, without a helper:
```go
func (o *evmTxStore) TransactionsWithAttempts(selector TransactionsWithAttemptsSelector) (txs []Tx, count int, err error) {
	query := " FROM evm.txes WHERE id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts)"
	var args []interface{}
	if selector.IdempotencyKey != nil {
		args = append(args, *selector.IdempotencyKey)
		// NOTE: we are lucky to have only one optional filter and WHERE is the root statement, so we can use AND 
		query += " AND idempotency_key = ?"
	}

	countQuery := "SELECT count(*)" + query
	if err = o.q.Get(&count, countQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("failed to exec count query: %w, sql: %s", err, countQuery)
	}

	args = append(args, selector.Limit, selector.Offset)
	selectQuery := "SELECT * " + query + " ORDER BY id desc LIMIT ? OFFSET ?"
	var dbTxs []DbEthTx
	if err = o.q.Select(&dbTxs, selectQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("failed to exec select query: %w, sql: %s", err, selectQuery)
	}
	txs = dbEthTxsToEvmEthTxs(dbTxs)
	err = o.preloadTxAttempts(txs)
	return
}
```

Reply:
Postgres uses numeric place holders for query arguments, so instead of `?` we need to use `$1, $2, etc.`. So on the line 
`selectQuery += " ORDER BY id desc LIMIT ? OFFSET ?"`  we do not know which placeholders to use , as idempotency key is optional

Reply:
Ah OK. So then is this correct? Is there any way to simplify it more? Maybe a `Sprintf` would be easier to read?
```go
func (o *evmTxStore) TransactionsWithAttempts(selector TransactionsWithAttemptsSelector) (txs []Tx, count int, err error) {
	query := " FROM evm.txes WHERE id IN (SELECT DISTINCT eth_tx_id FROM evm.tx_attempts)"
	var args []interface{}
	if selector.IdempotencyKey != nil {
		args = append(args, *selector.IdempotencyKey)
		// NOTE: we are lucky to have only one optional filter and WHERE is the root statement, so we can use AND 
		query += " AND idempotency_key = $1"
	}

	countQuery := "SELECT count(*)" + query
	if err = o.q.Get(&count, countQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("failed to exec count query: %w, sql: %s", err, countQuery)
	}

	selectQuery := "SELECT * " + query + " ORDER BY id desc "
	if len(args) == 0 {
		selectQuery += "LIMIT $1 OFFSET $2"
	} else {
		selectQuery += "LIMIT $2 OFFSET $3"
	}
	args = append(args, selector.Limit, selector.Offset)
	var dbTxs []DbEthTx
	if err = o.q.Select(&dbTxs, selectQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("failed to exec select query: %w, sql: %s", err, selectQuery)
	}
	txs = dbEthTxsToEvmEthTxs(dbTxs)
	err = o.preloadTxAttempts(txs)
	return
}
```

Reply:
Looks good, but I'm worried about this part
```go
if len(args) == 0 {
		selectQuery += "LIMIT $1 OFFSET $2"
} else {
		selectQuery += "LIMIT $2 OFFSET $3"
}
```
What if we need to add another optional parameter? It does not seem to be easy to extend. 
Why do you prefer this to the placeholder replacement?

Reply:
I'm not sure. @samsondav WDYT?

Reply:
sqlx comes with a helper to replace `?` with `$x`. Does this solve your problem? https://github.com/jmoiron/sqlx/blob/master/bind.go#L60